Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIStorageCocoa.mm

```diff
@@ -0,0 +1,143 @@
+{
+    NSString *errorMessage;
+    if (!isValidStorageRequest(webPageProxyIdentifier, [NSString stringWithFormat:@"%@.set()", (NSString *)toAPIPrefixString(storageType)], &errorMessage)) {
+        completionHandler(errorMessage);
+        return;
+    }
+
+    // FIXME: <https://webkit.org/b/267649> Set values for StorageAPISQLiteStore.
+    completionHandler(std::nullopt);
+}
+
+void WebExtensionContext::storageRemove(WebPageProxyIdentifier webPageProxyIdentifier, WebExtensionStorageType storageType, const Vector<String>& keys, CompletionHandler<void(ErrorString)>&& completionHandler)
+{
+    NSString *errorMessage;
+    if (!isValidStorageRequest(webPageProxyIdentifier, [NSString stringWithFormat:@"%@.remove()", (NSString *)toAPIPrefixString(storageType)], &errorMessage)) {
+        completionHandler(errorMessage);
+        return;
+    }
+
+    // FIXME: <https://webkit.org/b/267649> Remove values from StorageAPISQLiteStore.
+    completionHandler(std::nullopt);
+}
+
+void WebExtensionContext::storageClear(WebPageProxyIdentifier webPageProxyIdentifier, WebExtensionStorageType storageType, CompletionHandler<void(ErrorString)>&& completionHandler)
+{
+    NSString *errorMessage;
+    if (!isValidStorageRequest(webPageProxyIdentifier, [NSString stringWithFormat:@"%@.clear()", (NSString *)toAPIPrefixString(storageType)], &errorMessage)) {
+        completionHandler(errorMessage);
+        return;
+    }
+
+    // FIXME: <https://webkit.org/b/267649> Clear values in StorageAPISQLiteStore.
+    completionHandler(std::nullopt);
+}
+
+void WebExtensionContext::storageSetAccessLevel(WebPageProxyIdentifier webPageProxyIdentifier, WebExtensionStorageType storageType, const String& accessLevel, CompletionHandler<void(ErrorString)>&& completionHandler)
+{
+    static NSString * const callingAPIName = @"browser.session.setAccessLevel()";
+
+    NSString *errorMessage;
+    if (!isValidStorageRequest(webPageProxyIdentifier, callingAPIName, &errorMessage)) {
+        completionHandler(errorMessage);
+        return;
+    }
+
+    setSessionStorageAllowedInContentScripts([(NSString *)accessLevel isEqualToString:accessLevelTrustedAndUntrustedContexts]);
+
+    completionHandler(std::nullopt);
+}
+
+bool WebExtensionContext::isValidStorageRequest(WebPageProxyIdentifier webPageProxyIdentifier, NSString *callingAPIName, NSString **errorMessage)
+{
+    if (isBackgroundWebView(webPageProxyIdentifier))
+        return true;
+
+    RefPtr tab = getTab(webPageProxyIdentifier);
+    if (!tab) {
+        *errorMessage = toErrorString(callingAPIName, nil, @"tab not found");
+        return false;
+    }
+
+    auto *webView = tab->mainWebView();
+    NSURL *url = webView.URL;
+
+    if (isURLForThisExtension(url))
+        return true;
+
+    if (!hasPermission(url)) {
+        *errorMessage = toErrorString(callingAPIName, nil, @"this extension does not have access to this tab");
+        return false;
+    }
+
+    return true;
```

Comment:
This will fail for popup pages. I think the check can be simplified to focus on tab pages. If it isn't a tab page, just allow it since it is likely the background or a popup.

Also having the error be "tab" related is not correct here, since none of the storage APIs take tab ids. The check is about if they have access or not. The extensions can't get here if it does not have access to the tab, otherwise they couldn't have injected a script using `storage`, so I'm not sure we need a check here at all.

If we do need a check, it would be more like this:

```
    RefPtr tab = getTab(webPageProxyIdentifier, std::nullopt, WebExtensionContext::IgnoreExtensionAccess::No);
    if (!tab)
        return true;

    if (!tab->extensionHasAccess() || !tab->extensionHasPermission()) {
        *errorMessage = @"storage access not allowed";
        return false;
    }

    return true;
```

Reply:
Revised after talking about this:

```
RefPtr page = WebProcessProxy::webPage(webPageProxyIdentifier);
if (page && isURLForThisExtension(page->pageLoadState().activeURL()))
    return true;

RefPtr tab = getTab(webPageProxyIdentifier, std::nullopt, WebExtensionContext::IgnoreExtensionAccess::No);
if (!tab || !tab->extensionHasAccess() || !tab->extensionHasPermission()) {
    *errorMessage = @"storage access not allowed";
    return false;
}

return true;
```